Add python stack tracing option on on-demand flow

slgong-fb opened the pull request:
Summary:
Changes:
1. add an option in Config; can use 'PYTHON_STACK_TRACE=true' option (via .conf)
2. deliver PYTHON_STACK_TRACE value to kineto_client_interface start()
3. abstract class also changed.

Test Plan:
1. launch a python test case with the following command for on-demand flow:
echo -e "PYTHON_STACK_TRACE=true" > /tmp/scott_kineto.conf && dyno gputrace --gputrace_duration 300ms --gpuconf /tmp/scott_kineto.conf

2. Then, we can see with_stack enabled as intended from output log:
INFO:2022-06-27 15:00:16 1009443:1011716 kineto_client_interface.cpp:22] withStack : 1

Differential Revision: D37410204

robieta commented on torch/csrc/profiler/kineto_client_interface.cpp:
```diff
@@ -18,16 +18,19 @@ class LibKinetoClient : public libkineto::ClientInterface {
+        /*with_modules=*/false};
     std::set<ActivityType> activities{ActivityType::CPU};
-    auto scopes = {at::RecordScope::FUNCTION, at::RecordScope::USER_SCOPE};
+    std::unordered_set<at::RecordScope> scopes;
```
Why not just
```
auto scopes = {
  at::RecordScope::FUNCTION,
  at::RecordScope::USER_SCOPE, 
  at::RecordScope::BACKWARD_FUNCTION};
```

robieta replied:
Bumping this comment.

robieta commented on torch/csrc/autograd/profiler_python.cpp:
```diff
@@ -301,7 +301,6 @@ void ValueCache::trimPrefixes() {
   static auto prefixes = py::module::import("torch.profiler.python_tracer")
                              .attr("_prefix_regex")()
                              .cast<std::vector<std::string>>();
-
```
Nit: remove to keep the file history clean.

robieta commented on torch/csrc/profiler/kineto_client_interface.cpp:
```diff
@@ -18,16 +18,29 @@ class LibKinetoClient : public libkineto::ClientInterface {
     reportInputShapes_ = setupOpInputsCollection;
   }
 
+#ifdef USE_KINETO_UPDATED
```
We should try to minimize the use of `#ifdef`, especially when it alters a function signature. In `kineto_shim.h` we convert these `#ifdef USE_KINETO` into a constexpr bool (https://github.com/pytorch/pytorch/blob/master/torch/csrc/profiler/kineto_shim.h#L32). I think you should do the same with `USE_KINETO_UPDATED` and then use a more C++ idiomatic way of guarding. Do we care about compile time vs. runtime check? If the latter is enough you can just do something like `TORCH_CHECK(!withStack || kUseKinetoUpdated);`. If you want compile time check you'll have to do something like using a template to make `withStack` constexpr.

robieta reviewed (approved):
Approving, with an action item to clean up the interface in a follow up PR.

slgong-fb committed:
Add python stack tracing option on on-demand flow (#80919)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/80919

X-link: https://github.com/pytorch/kineto/pull/628

Changes:
1. add an option in Config; can use 'PYTHON_STACK_TRACE=true' option (via .conf)
2. deliver PYTHON_STACK_TRACE value to kineto_client_interface start()
3. abstract class also changed.

Trace after changes by running //kineto/libkineto/fb/integration_tests/trace_tester.cpp (requested by chaekit)
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree%2Ftraces%2Fdynocli%2F0%2F1657304871%2F127.0.0.1%2Flibkineto_activities_3502962.json.gz&bucket=gpu_traces

Test Plan:
1. launch a python test case with the following command for on-demand flow:
echo -e "PYTHON_STACK_TRACE=true" > /tmp/scott_kineto.conf && dyno gputrace --gputrace_duration 300ms --gpuconf /tmp/scott_kineto.conf

2. Then, we can see with_stack enabled as intended from output log:
INFO:2022-06-27 15:00:16 1009443:1011716 kineto_client_interface.cpp:22] withStack : 1

Reviewed By: chaekit

Differential Revision: D37410204

fbshipit-source-id: 7962c05b3bccb1359dfc46d832abda10caeb0e81

janeyx99 commented:
@pytorchbot revert -m "Sorry, reverting as this broke buck build/test https://hud.pytorch.org/pytorch/pytorch/commit/f50a248a5eacb9a9aa475a9e610486aea136e4f5" -c nosignal

Could you please add ciflow/periodic to the reland PR to capture the signal?

kit1980 commented:
> @pytorchbot revert -m "Sorry, reverting as this broke buck build/test https://hud.pytorch.org/pytorch/pytorch/commit/f50a248a5eacb9a9aa475a9e610486aea136e4f5" -c nosignal

To be clear, buck build helped to caught an actual issue with the code: `void start(bool withStack) override` has one bool arg, while the parent's `start` has no arguments, so you can't override.

chaekit commented:
the buck build failure is legit but not sure why `linux-bionic-py3.7-clang9`

```
ERROR [0.001s]: test_bce_with_logits_has_correct_forward_grad (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 9557, in test_bce_with_logits_has_correct_forward_grad
    (output, target), check_forward_ad=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 3148, in gradcheck
    return torch.autograd.gradcheck(fn, inputs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1418, in gradcheck
    return _gradcheck_helper(**args)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1427, in _gradcheck_helper
    func_out = func(*tupled_inputs)
  File "test_nn.py", line 9556, in <lambda>
    gradcheck(lambda self, target: nn.BCEWithLogitsLoss(reduction=reduction)(self, target),
  File "test_nn.py", line 9556, in <lambda>
    gradcheck(lambda self, target: nn.BCEWithLogitsLoss(reduction=reduction)(self, target),
  File "/var/lib/jenkins/torchdynamo/torchdynamo/eval_frame.py", line 93, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 655, in call_wrapped
    return self._wrapped_call(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 277, in __call__
    raise e
  File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 267, in __call__
    return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1186, in _call_impl
    return forward_call(*input, **kwargs)
TypeError: forward() takes 2 positional arguments but 3 were given
```

this is strictly profiler change and does not touch any PyTorch core. also it passed in CI before landing

robieta commented:
> the buck build failure is legit but not sure why `linux-bionic-py3.7-clang9`
> 
> ```
> ERROR [0.001s]: test_bce_with_logits_has_correct_forward_grad (__main__.TestNN)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "test_nn.py", line 9557, in test_bce_with_logits_has_correct_forward_grad
>     (output, target), check_forward_ad=True)
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 3148, in gradcheck
>     return torch.autograd.gradcheck(fn, inputs, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1418, in gradcheck
>     return _gradcheck_helper(**args)
>   File "/opt/conda/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 1427, in _gradcheck_helper
>     func_out = func(*tupled_inputs)
>   File "test_nn.py", line 9556, in <lambda>
>     gradcheck(lambda self, target: nn.BCEWithLogitsLoss(reduction=reduction)(self, target),
>   File "test_nn.py", line 9556, in <lambda>
>     gradcheck(lambda self, target: nn.BCEWithLogitsLoss(reduction=reduction)(self, target),
>   File "/var/lib/jenkins/torchdynamo/torchdynamo/eval_frame.py", line 93, in _fn
>     return fn(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 655, in call_wrapped
>     return self._wrapped_call(self, *args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 277, in __call__
>     raise e
>   File "/opt/conda/lib/python3.7/site-packages/torch/fx/graph_module.py", line 267, in __call__
>     return super(self.cls, obj).__call__(*args, **kwargs)  # type: ignore[misc]
>   File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1186, in _call_impl
>     return forward_call(*input, **kwargs)
> TypeError: forward() takes 2 positional arguments but 3 were given
> ```
> 
> this is strictly profiler change and does not touch any PyTorch core. also it passed in CI before landing

It's unrelated to the BCE test. The issue is that before the class hierarchy looked something like:
```
struct base {
  virtual void foo();
};

struct derived {
  void foo() override;
};
```

And it got changed to:

```
struct base {
  virtual void foo();
};

struct derived {
  // Not a legitimate override.
  void foo(bool) override;
};
```

Honestly, I think the question isn't why did buck fail but rather "why didn't all other CI fail"? (My suspicion is that PyTorch is using Kineto's CMake files so whatever lets it through Kineto CI also lets it through PyTorch CI. But just a hunch.) It didn't fail CI because buck CI doesn't run on PRs, it runs periodically. If it was an issue with buck I would have fought the revert (I have also been bitten by this and it's very annoying), but in this case buck was flagging a real problem as @kit1980 points out.

Speaking of which, this seems like a good reason to put buck build in the per PR group.
